discovery/cvm/cvm.go

```diff
@@ -0,0 +1,305 @@
+	Endpoint        string         `yaml:"endpoint,omitempty"`
+	Region          string         `yaml:"region,omitempty"`
+	Port            int            `yaml:"port"`
+	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
+	Filters         []*Filter      `yaml:"filters"`
+
+	// Tencent CVM Auth: https://github.com/tencentcloud/tencentcloud-sdk-go
+	SecretId  string `yaml:"secret_id,omitempty"`  // env set PROMETHEUS_DS_CVM_SECRET_ID
+	SecretKey string `yaml:"secret_key,omitempty"` // env set PROMETHEUS_DS_CVM_SECRET_KEY
+	Token     string `yaml:"token,omitempty"`      // env set PROMETHEUS_DS_CVM_TOKEN
+	RoleArn   string `yaml:"role_arn,omitempty"`   // env set PROMETHEUS_DS_CVM_ROLE_ARN
+}
+
+// Discovery periodically performs CVM-DescribeInstance requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	conf   *SDConfig
+	logger log.Logger
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+
+	if c.Region == "" {
+		// todo get region from metadata api if region is blank
+		return errors.New("CVM SD configuration requires a region")
+	}
+	for _, f := range c.Filters {
+		if len(f.Values) == 0 {
+			return errors.New("CVM SD configuration filter values cannot be empty")
+		}
+	}
+	return nil
+}
+
+// NewDiscovery returns a new cvmDiscovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) *Discovery {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	d := &Discovery{
+		conf:   conf,
+		logger: logger,
+	}
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"cvm",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	_ = level.Debug(d.logger).Log("refresh with sd conf:", d.conf.Region, d.conf.Endpoint, d.conf.Filters)
+
+	client, err := getCvmClient(d.conf)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create cvm client")
+	}
+	tg := &targetgroup.Group{
+		Source: d.conf.Region,
+	}
+
+	var filters []*cvm.Filter
+	for _, f := range d.conf.Filters {
+		filters = append(filters, &cvm.Filter{
+			Name:   common.StringPtr(f.Name),
+			Values: common.StringPtrs(f.Values),
+		})
+	}
+	request := cvm.NewDescribeInstancesRequest()
+	request.Filters = filters
+	request.Limit = common.Int64Ptr(DefaultPageLimit)
+
+	var offset int64 = 0
+
+	for {
+		request.Offset = common.Int64Ptr(offset)
+
+		_ = level.Debug(d.logger).Log("retrieving instances, offset:", *request.Offset)
+
+		resp, err := client.DescribeInstances(request)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not describe cvm instances")
+		}
+
+		for _, inst := range resp.Response.InstanceSet {
+			if inst.PrivateIpAddresses == nil {
+				continue
```

Comment:
in which case do we have that?

Reply:
in fact we never have that，this is just defensive programming

Reply:
then you should check for len = 0  too ?

Reply:
yes, fixed, check len = 0  is the right solution.